### DISCUSSION TITLE
Stealth 4 All: New Visibility Modifier for all 4 modes!

### DISCUSSION THREAD

**Discussion by X-Funkin** [ppy/osu#19289] [Ideas]
Inspired by _Dance Dance: Revolution_, I decided to add _Stealth_ into osu!
here it is in osu!mania:

https://user-images.githubusercontent.com/84550797/180340142-e607ee73-eafc-4884-8a98-dd2952ebb6e1.mp4

an actual player would look like this:

https://user-images.githubusercontent.com/84550797/180340187-2a8a0d90-ed00-4b8d-936c-55db29e26dbe.mp4

Now, it does _technically_ work in all 4 game modes, as seen here:

https://user-images.githubusercontent.com/84550797/180340676-bafef432-3d2e-45d1-bbe6-5a31645545dc.mp4

But there are a couple of problems that I can't figure out:
**Problem in Taiko**
![osu_2022-07-21_20-51-35](https://user-images.githubusercontent.com/84550797/180340807-2437f941-1243-4f04-b552-bba9c416532b.png)

As we can see, the drum roll hits for when you hit the drum rolls are _not_ invisible, and this is an issue because `drumRollHitContainer` is a private variable, easy enough to fix, right? Well, the ModStealth.cs i made uses `public virtual void Update(Playfield playfield)` function like the no-scope modifier and i can't convert it to a TaikoPlayfield with `var taikoPlayfield = (TaikoPlayfield) playfield` like i could with mania, it just crashes because it can't convert `BarLinePlayfeild` into `TaikoPlayfeild`

**Problem in Standard**
So osu!stealth works flawlessly, except for the "Increase Visibility of first object with visibility mods" option that's present in the game.
Hidden and Traceable (and probably others i don't know) make use of that option, but I cannot for the life of me figure out a good way to have that option work with the stealth modifier that doesn't include calling a function multiple times possibly for each object possibly for each frame just to change a number.  Currently the modifier works by updating the `playfield.HitObjectsContainer.Alpha`, since all the visibilities of all objects are completely synced, i don't have to change the visibility of each individually. But right now i'm stuck on standard and taiko

Mania, tho, is pretty much completely done, all i have to do is update it a bit and pass the continual integration tests, then it'll be ready for a code review and PR
[Here's](https://github.com/X-Funkin/funky-osu/tree/osu-mania-stealth-mod) the branch for the Mania stealth mod, pretty much ready
And [here's](https://github.com/X-Funkin/funky-osu/tree/osu-Stealth-4-all) the branch with all 4 modes, but both problems present

**Comment by frenzibyte** [comment]
Interesting work on this, though I'm not entirely sure if this is any different to the flashlight mod, or any mod on other rulesets which covers the playfield and uncovers every time you miss.

**Reply by X-Funkin** [reply]
Mostly with the current visibility mods, you don't have to completely memorize the charts to FC, like you still visually see which notes to hit in modes like mania and taiko, for instance, only restricted

this mod, tho, makes skills like memorization much more important.

It's extremely niche, i'll admit, but alot of the other new mods in standard are also extremely niche, too, so i think it fits in

but also
_it's compatible with NoScope_

https://user-images.githubusercontent.com/84550797/180361506-ccc38b76-3af2-4a51-bdc7-d9e12176eea1.mp4


You can click nothing with nothing and i think that's kinda funny, especially since there will definitely be atleast _one_ player in osu's entire playerbase who would try and specialize in something like this

but yeah, in all honesty if mods like NoScope and a bunch of the movement and "fun" mods weren't a thing, i probably wouldn't have added Stealth

Thank you for the reply!!
